core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncPathCache.java

```diff
@@ -11,201 +11,348 @@
  */
 @ThreadSafe
-public final class UfsSyncPathCache {
+public class UfsSyncPathCache {
   private static final Logger LOG = LoggerFactory.getLogger(UfsSyncPathCache.class);
 
-  /** Number of paths to cache. */
-  private static final int MAX_PATHS =
-      Configuration.getInt(PropertyKey.MASTER_UFS_PATH_CACHE_CAPACITY);
-
-  /** Cache of paths which have been synced. */
-  private final Cache<String, SyncTime> mCache;
-
+  public final Cache<String, SyncState> mItems;
   private final Clock mClock;
+  // we keep the root as a separate value as we never want to evict it,
+  // and instead we just update it under a lock
+  private final SyncState mRoot = new SyncState(false);
+  private final Lock mRootLock = new ReentrantLock();
 
   /**
    * Creates a new instance of {@link UfsSyncPathCache}.
    * @param clock the clock to use to compute sync times
    */
   public UfsSyncPathCache(Clock clock) {
+    this(clock, null);
+  }
+
+  @VisibleForTesting
+  UfsSyncPathCache(Clock clock,
+                   @Nullable BiConsumer<String, SyncState> onRemoval) {
     mClock = clock;
-    mCache = CacheBuilder.newBuilder().maximumSize(MAX_PATHS).build();
+    mItems = CacheBuilder.newBuilder()
+        .removalListener(
+            (removal) -> {
+              if (removal.wasEvicted() && removal.getKey() != null && removal.getValue() != null) {
+                if (onRemoval != null) {
+                  onRemoval.accept((String) removal.getKey(), (SyncState) removal.getValue());
+                }
+                onCacheEviction((String) removal.getKey(), (SyncState) removal.getValue());
+              }
+            })
+        .maximumSize(Configuration.getInt(
+            PropertyKey.MASTER_UFS_PATH_CACHE_CAPACITY))
+        .build();
   }
 
   /**
-   * Notifies the cache that the path was synced.
-   *
-   * @param path the path that was synced
-   * @param descendantType the descendant type that the path was synced with
-   * @param syncTime the time to set the sync success to, if null then the current
-   *                 clock time is used
+   * Called when an element is evicted from the cache.
+   * @param path the path
+   * @param state the state
    */
-  public void notifySyncedPath(
-      String path, DescendantType descendantType, @Nullable Long syncTime) {
-    long syncTimeMs = syncTime == null ? mClock.millis() :
-        syncTime;
-    mCache.asMap().compute(path, (key, oldSyncTime) -> {
-      if (oldSyncTime != null) {
-        // update the existing sync time
-        oldSyncTime.updateSync(syncTimeMs, descendantType);
-        return oldSyncTime;
-      }
-      return new SyncTime(syncTimeMs, descendantType);
-    });
+  void onCacheEviction(String path, SyncState state) {
+    try {
+      // on eviction, we must mark our parent as needing a sync with our invalidation time
+      notifyInvalidationInternal(PathUtils.getParent(path), state.mInvalidationTime);
```

Comment:
if a directory has many children then does this mean that directory is likely to be continuously invalidated? In fact I guess the cache is LRU-based so the chance of eviction is more likely to have a pattern and is more linear to how many files you have.

In the extreme case, if the size of a directory is larger than you cache, then this dir will always stay invalidated and trigger recursive ls to the UFS, i guess that's NOT true? I don't immediately see how that's avoided and could you explain that part in the comments? I figure that'd be very helpful to whoever's reading it.

Reply:
A sync will only be needed if the invalidation time is greater than the most recent validation time. So if a sync was performed recently, but a child that was invalidated long ago is evicted, then it still will not need a sync.

The LRU cache should evict the files before the directory, since each time we process a path, either sync or invalidation, the directories from the root are accessed.

I added some more comments.

Reply:
Ah thanks for the explanation.

I guess another fact we had to accept is, if a directory constantly has files created/deleted, the directory can be always invalidated? Then we have to hope there are not so many direct children under it. Is my guess true? (I don't expect that to be handled right away)

Reply:
Yes if there are frequent listings during invalidations of the files internally then the listings will check with the UFS.